### DISCUSSION TITLE
[Workflow] What is the best way to display the reason a place cannot be reached

### DISCUSSION THREAD

**Discussion by VincentLanglet** [symfony/symfony#57496] [Q&A] [answered]
Hi, @lyrixx

You introduced the TransitionBlockerList and I have small question/feature request about it 
https://github.com/symfony/symfony/blob/c852b8bff169669175489bb468c8c8328ecaae53/src/Symfony/Component/Workflow/TransitionBlockerList.php#L21

If we want to know if we can reach a place, there is a method for this
```
$this->workflow->can($subject, $place);
```
but if I want to know the reason I cannot reach the place I need to play with TransitionBlockerList, isn't it 
```
$this->workflow->buildTransitionBlockerList($subject, $place);
```

But then it seems to me there is no easy way to stringify this this list:
- I don't have a method
```
$blockerList->toString()
```
- I cannot write something like 
```
$blockerList->get(0)->getMessage()
```
- I cannot technically write (since getIterator returns an iterable the implementation might change one day so I cannot rely on the fact it's an ArrayIterator)
```
$blockerList->getIterator()[0]->getMessage()
```

What do you think about one of the options:
- Adding a toString method on TransitionBlockerList (and on TransitionBlocker ?) ; in the same way it exist on ConstraintViolationListInterface and ConstraintViolationInterface.
- Adding a `get()` method on TransitionBlockerList
- Changing the return type of `getIterator` for a more precise one

**Comment by lyrixx** [comment, marked as answer]
Hello, you have [an example here](https://github.com/lyrixx/SFLive-Paris2016-Workflow/blob/master/templates/article/show.html.twig#L52-L68)

Anyway, you can do:

```php
$errors = $this->workflow->buildTransitionBlockerList($subject, $place);
foreach($errors as $error) {
 ...
}
```

And if you really need the first value, you can break the foreach loop avec the first iteration, or:

```php
$errors = $this->workflow->buildTransitionBlockerList($subject, $place);
$error0 = array_values(iterator_to_array($errors))[0] ?? null;
```

o

**Reply by VincentLanglet** [reply]
Thanks

I end up doing 
```
$errors = $this->workflow->buildTransitionBlockerList($subject, $place);
if ($errors->isEmpty()) {
    return null; // Because using `can` and `buildTransitionBlockerList` would compute the transition blocker list two times.
}
$error0 = array_values(iterator_to_array($errors))[0] ?? null;
```
But I would love a native method to avoid `array_values(iterator_to_array($errors))[0]`.

Anyway, it does the job. Have a good day.

**Comment by stof** [comment]
I don't think we will change the return type of `getIterator` as this would prevent us to perform any refactoring of the implementation using a different kind of iterator.

The current API has a benefit though: it encourages dealing with all blockers instead of only one of them.

**Reply by VincentLanglet** [reply]
>The current API has a benefit though: it encourages dealing with all blockers instead of only one of them.

That's also why I proposed another solution which is to provide a toString method on TransitionBlockerList (and on TransitionBlocker ?) ; in the same way it exist on ConstraintViolationListInterface and ConstraintViolationInterface.

This would simplify the way we have to deal with all blockers just to generate a string.
